Move `pkg/certificates` from `control-protocol` to `networking`

ReToCode opened the pull request:
# Motivation
- Currently Serving depends on `control-protocol` just for `pkg/certificates`
- The constants and parts of the code in `control-protocol` is needed in `net-*` implementations

# Changes
- Move `pkg/certificates` from `control-protocol` to `networking`
- Update the dependencies accordingly
- Fixed lint warnings

# Testing
- Testing in https://github.com/knative/serving/pull/13938.

**Release Note**
None

**Docs**
None

/kind cleanup

/cc @evankanderson, @nak3, @davidhadas, @dprotaso 

Fixes #805

ReToCode commented:
I'd like to ignore lint errors to keep the code the same as in the origin and fix those in a new PR afterwards.

```
  Error: G101: Potential hardcoded credentials (gosec)
  Error: G101: Potential hardcoded credentials (gosec)
  Warning: var-naming: const LegacyFakeDnsName should be LegacyFakeDNSName (revive)
  Warning: var-naming: const FakeDnsName should be FakeDNSName (revive)
  Warning: var-naming: func parameter routingId should be routingID (revive)
  Warning: var-naming: func parameter routingId should be routingID (revive)
  Warning: var-naming: var routingId should be routingID (revive)
  Warning: var-naming: const secretRoutingId should be secretRoutingID (revive)
  Error: createCert - result cert is never used (unparam)
  Error: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
  Error: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
```

Should be fine, as we have to update all usages anyway. FYI @davidhadas, @nak3

nak3 commented:
/lgtm
/approve
/hold for @davidhadas or other reviewers.

davidhadas commented:
Would it not be better to fix the lint and other issues in the original `control-protocol` repo and only then move to  `networking`?
I think it is cleaner. 
Consider any other PR that will be needed in `networking` until it is fixed - it will fail on Lint.

ReToCode commented:
This would be way more work (as this here is already done and tested) for something that is removed in `control-protocol`, also it seems that control-protocol has different lint settings, otherwise it would have failed there as well. But I get your point about the builds, then I'd prefer to go with fixing those in this PR in a separate commit.

Regarding the symbols. Agreed in general, but these are all just recently introduced by you for an undocumented alpha feature. We should be fine with changing them, right?

- LegacyFakeDnsName -> just recently renamed
- FakeDnsName -> just recently introduced
- routingID -> internal
- secretRoutingID -> internal

davidhadas commented:
> This would be way more work (as this here is already done and tested) for something that is removed in control-protocol, also it seems that control-protocol has different lint settings, otherwise it would have failed there as well. But I get your point about the builds, then I'd prefer to go with fixing those in this PR in a separate commit.

ok

> Regarding the symbols. Agreed in general, but these are all just recently introduced by you for an undocumented alpha feature. We should be fine with changing them, right?

>LegacyFakeDnsName -> just recently renamed
FakeDnsName -> just recently introduced
routingID -> internal
secretRoutingID -> internal

afaik, `FakeDnsName` is being used - this is why I did not create or change it. All I did is deprecate  it and replaced it with LegacyFakeDnsName to make people notice that we are making changes and moving away from the FakeDnsName.

ReToCode commented:
/hold as I need to check if test failures in Serving are related or flakes.

evankanderson reviewed (commented):
Apparently, I never sent my review.

If you want to clean these in the future, I don't feel very strongly about adding more to this PR.

ReToCode commented:
> If you want to clean these in the future, I don't feel very strongly about adding more to this PR.

@evankanderson I think I'd prefer to merge this one as closely to the original code from `control-protocol` and do a [follow-up](https://github.com/knative/networking/issues/805) on the improvements. We also need to change all the usages of the code in serving and all net-* repos, then we'd also have better feedback if we break something in the dependant repos.

ReToCode commented:
/unhold

@evankanderson [gentle ping](https://github.com/knative/networking/pull/802#issuecomment-1535705966)

ReToCode commented:
Rebased here, rebased & retriggered the [checking PR](https://github.com/knative/serving/pull/13938).

@evankanderson can you please take a look? ^^

dprotaso commented on pkg/certificates/certs.go:
```diff
@@ -0,0 +1,196 @@
+	logger := logging.FromContext(ctx)
+	caKeyPair, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		logger.Errorw("error generating random key", zap.Error(err))
```
+1

dprotaso commented on pkg/certificates/constants.go:
```diff
@@ -0,0 +1,59 @@
+const (
+	Organization = "knative.dev"
+	//nolint:all
+	LegacyFakeDnsName = "data-plane." + Organization
+	//nolint:all
+	FakeDnsName            = LegacyFakeDnsName // Deprecated
```
Do we need the deprecated constants here?

dprotaso commented on pkg/certificates/certs.go:
```diff
@@ -0,0 +1,196 @@
+
+// ValidateCert checks the expiration of the certificate
+func ValidateCert(cert *x509.Certificate, rotationThreshold time.Duration) error {
+	if !cert.NotAfter.After(time.Now().Add(rotationThreshold)) {
```
or maybe
`time.Now().Add(threshold).After(cert.NotAfter)` ?

ReToCode commented:
[Initially](https://github.com/knative/networking/pull/802#issuecomment-1535705966) I intended to keep the code the same as in the source, but I'm also fine with fixing it in this PR.

dprotaso commented:
/lgtm

nak3 commented:
A few lint error happens. Will we fix another PR or fix here?

nak3 commented:
/lgtm
/hold 

Thank you! `/hold` for test.
